spec/shrinkwrap-and-lock.md

```diff
@@ -0,0 +1,273 @@
+will be updated.
+
+If neither exist then a `package-lock.json` should be generated.
+
+If a `package.json` does not exist, it should be generated.  The generated
+`package.json` should be empty, as in:
+
+```
+{
+   "dependencies": {
+   }
+}
+```
+
+If the user wants to get a default package name/version added they can run `npm init`.
+
+### `npm shrinkwrap`
+
+If a `package-lock.json` exists, rename it to `npm-shrinkwrap.json`.
+Refresh the data from the installer's ideal tree.
+
+The top level `name` and `version` come from the `package.json`.  It is an
+error if either are missing or invalid.
+
+`created-with` should be filled in from `npm` itself.
+
+#### dependencies.dev
+
+This is `true` if this dependency is ONLY installed to fulfill either a top
+level development dependency, or one of its transitive dependencies. 
+
+Given:
+```
+B (Dev) → C
+```
+
+Then both B and C would be `dev: true`.
+
+Given:
+```
+A → B → C
+B (Dev) -> C
+```
+
+Then all dependencies would be `dev: false`.
+
+#### dependencies.optional
+
+This is `true` if this dependency is ONLY ever either an optional dependency
+or a transitive dependency of optional dependencies.
+
+Given:
+```
+A (Opt) → B → C
+```
+
+Then all three of A, B and C would be flagged as optional.
+
+Given:
+```
+A (Opt) → B → C
+D → C
+```
+
+Then A and B would be flagged as optional, but C would not be.
+
+Given:
+```
+A (Opt) → B → C
+D → A
+```
+
+Then none would be flagged as optional.
+
+## Installing
+
+The `package-lock.json` or `npm-shrinkwrap.json` describes the exact tree
+that `npm` should create.  Any deviation between the `package.json` and the
+shrinkwrap/lock should result in a warning be issued.  This includes:
+
+* Modules in `package.json` but missing from `npm-shrinkwrap.json` or `package-lock.json`
+* Modules in the `npm-shrinkwrap.json` or `package-lock.json` but missing from the `package.json`.
+* Modules in `package.json` whose specifiers don't match the version in `npm-shrinkwrap.json` or `package-lock.json`.
+
+Warn if the `shrinkwrap-version` in the `npm-shrinkwrap.json` is for a different
+major version than we implement.
+
+Module resolution from shrinkwrap data works as such:
+
+* If install was run with `--resolve-links` and a dependency has a `link`
+  property then a symlink is made using that.  If the version of the
+  destination can not be matched to the shrinkwrap and/or the package.json
+  then a warning will be issued.
+
+* Otherwise, if a `integrity` is available then we try to install it from the cache using it.
+
+If `integrity` is unavailable or we are unable to locate a module from the `integrity` then:
+
+* If `shrinkwrap-version` is set:
+  * Install using the value of `version` and validate the result against the
+    `integrity`.
+* Otherwise, try these in turn and validate the result against the `integrity`:
+  * `resolved`, then `from`, then `version.
+  * `from` can be either `package@specifier` or just `specifier`.
```

Comment:
It could, but also this would mean you get warnings for published npm packages that publish an `npm-shrinkwrap.json`. I don't feel like it has that much value if it will be adapted next time someone runs a new install.